File: 0038_count_and_say.py
# https://leetcode-cn.com/problems/count-and-say/

class Solution:
    def countAndSay(self, n:int) -> str:
        if n == 1:
            return '1'
        
        res = []
        s = self.countAndSay(n-1)
        count = 1
        for i in range(len(s)-1):
            if s[i] == s[i+1]:
                count = count + 1
            else:
                res.append(str(count) + s[i])
                count = 1
        res.append(str(count) + s[-1])
        return res

File: Cpp/0038_count_and_say.cpp
// https://leetcode.cn/problems/count-and-say/

class Solution {
public:
    string countAndSay(int n) {
        if (n == 1) {
            return "1";
        }

        string res = "", res_last = countAndSay(n-1);
        char cur = res_last[0];
        int cnt = 1;
        for (int i = 1; i < res_last.size(); ++i) {
            if (cur == res_last[i]) {
                cnt = cnt + 1;
            }
            else {
                res = res + to_string(cnt) + cur;
                cur = res_last[i];
                cnt = 1;
            }
        }
        res = res + to_string(cnt) + cur;
        return res;
    }
};
